backend/cfg/cfg_dataflow.ml

```diff
@@ -130,5 +132,174 @@ module Forward (D : Domain) (T : Transfer with type domain = D.t) :
       update ~normal:true ~exn:false normal;
       update ~normal:false ~exn:true exceptional
     done;
-    if !iteration < max_iteration then Result.Ok res else Result.Error res
+    if WorkSet.is_empty !work_set then Result.Ok res else Result.Error res
+end
+
+module type Backward_domain = sig
+  type t
+
+  val bot : t
+
+  val compare : t -> t -> int
+
+  val join : t -> t -> t
+
+  val less_equal : t -> t -> bool
+
+  val to_string : t -> string
+end
+
+module type Backward_transfer = sig
+  type domain
+
+  val basic : domain -> exn:domain -> Cfg.basic Cfg.instruction -> domain
+
+  val terminator :
+    domain -> exn:domain -> Cfg.terminator Cfg.instruction -> domain
+
+  val exception_ : domain -> domain
+end
+
+module type Backward_S = sig
+  type domain
+
+  type map = domain Label.Tbl.t
+
+  val run :
+    Cfg.t -> ?max_iteration:int -> init:domain -> unit -> (map, map) Result.t
+end
+
+module Backward
+    (D : Backward_domain)
+    (T : Backward_transfer with type domain = D.t) :
+  Backward_S with type domain = D.t = struct
+  (* CR xclerc for xclerc: see what can be shared with `Forward`. *)
+
+  type domain = D.t
+
+  type map = domain Label.Tbl.t
+
+  module WorkSetElement = struct
+    type t =
+      { label : Label.t;
+        value : domain
+      }
+
+    let compare { label = left_label; value = left_value }
+        { label = right_label; value = right_value } =
+      match Label.compare left_label right_label with
+      | 0 -> D.compare left_value right_value
+      | res -> res
+  end
+
+  module WorkSet = Set.Make (WorkSetElement)
+
+  let transfer_block : domain -> exn:domain -> Cfg.basic_block -> domain =
+   fun value ~exn block ->
+    ListLabels.fold_right block.body
+      ~init:(T.terminator value ~exn block.terminator) ~f:(fun instr value ->
+        T.basic value ~exn instr)
+
+  let create : Cfg.t -> init:domain -> map * WorkSet.t ref =
+   fun cfg ~init ->
+    let map = Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks) in
+    let set = ref WorkSet.empty in
+    let value = init in
+    Cfg.iter_blocks cfg ~f:(fun label _block ->
+        Label.Tbl.replace map label value;
+        set := WorkSet.add { WorkSetElement.label; value } !set);
+    map, set
+
+  let remove_and_return :
+      Cfg.t -> WorkSet.t ref -> WorkSetElement.t * Cfg.basic_block =
+   fun cfg set ->
+    let element = WorkSet.choose !set in
+    set := WorkSet.remove element !set;
+    element, Cfg.get_block_exn cfg element.label
+
+  let run :
+      Cfg.t -> ?max_iteration:int -> init:domain -> unit -> (map, map) Result.t
+      =
+   fun cfg ?(max_iteration = max_int) ~init () ->
+    let res, work_set = create cfg ~init in
+    let iteration = ref 0 in
+    (* note: `handler_map` contains the value at the *start* of the block. *)
+    let handler_map : D.t Label.Tbl.t =
```

Comment:
The problem, as I understand it, is that when you
have just processed a given node, you will update
(some of) its regular predecessors with the "in"
value of the processed node. For the predecessors,
this "in" value is "joined" to compute their "out"
values. The problem with exceptional nodes is
that their "in" values must be "joined" not at the
end of the block, but at the point where an
exceptional edge may be followed. The situation
is slightly simpler now since a given block can
raise in at most one point (it was not the case
when this code was written, as far as I can
remember), but I think there are still two places
where the raise can occur: the last instruction
of the body and the terminator. The latter case
is indeed as simple to handle as for a regular
edge. The former still seem to ask for a different
treatment, unless perhaps there is an invariant
saying that when the last instruction is raising
then the terminator does not affect the value.
I can certainly see that property hold for the
current liveness analysis if _e.g._ we guarantee
that the terminator in such as case is necessarily
an `Always`.

Reply:
> The problem with exceptional nodes is
> that their "in" values must be "joined" not at the
> end of the block, but at the point where an
> exceptional edge may be followed.

Yes, the problem is that there are two such points now (terminator or last instruction).